File: cpp/assemble_vector.cpp
// Copyright (C) 2021 Jorgen S. Dokken & Nathan Sime
//
// This file is part of DOLFINX_MPC
//
// SPDX-License-Identifier:    LGPL-3.0-or-later

#include "assemble_vector.h"
#include <dolfinx/fem/Constant.h>
#include <dolfinx/fem/DirichletBC.h>
#include <dolfinx/fem/utils.h>
#include <iostream>

namespace
{

/// Given a local element vector, move all slave contributions to the global
/// (local to process) vector.
/// @param [in, out] b The global (local to process) vector
/// @param [in, out] b_local The local element vector
/// @param [in] b_local_copy Copy of the local element vector
/// @param [in] num_dofs The number of degrees of freedom in the local vector
/// @param [in] bs The element block size
/// @param [in] slave_indices Indices of slaves in the local element vector
/// (relative to the multi point constraint)
/// @param[in] local_indices The slave indices (local to process)
/// @param[in] masters Adjacency list with master dofs
/// @param[in] coeffs Adjacency list with the master coefficients
template <typename T>
void modify_mpc_vec(
    const xtl::span<T>& b, const xtl::span<T>& b_local,
    const xtl::span<T>& b_local_org, const int num_dofs, const int bs,
    const xtl::span<const int32_t>& slave_indices,
    const std::vector<std::int32_t>& local_indices,
    const std::shared_ptr<const dolfinx::graph::AdjacencyList<std::int32_t>>&
        masters,
    const std::shared_ptr<const dolfinx::graph::AdjacencyList<T>>& coeffs)
{

  // Given the set of slave indices in the cell, flatten all slaves, masters and
  // coefficients to equal length arrays for easy insertion
  std::vector<std::int32_t> flattened_masters;
  std::vector<std::int32_t> flattened_slaves;
  std::vector<std::int32_t> slaves_loc;
  std::vector<T> flattened_coeffs;
  for (std::int32_t i = 0; i < slave_indices.size(); ++i)
  {
    xtl::span<const std::int32_t> local_masters
        = masters->links(slave_indices[i]);
    xtl::span<const T> local_coeffs = coeffs->links(slave_indices[i]);

    for (std::int32_t j = 0; j < local_masters.size(); ++j)
    {
      slaves_loc.push_back(i);
      flattened_slaves.push_back(local_indices[i]);
      flattened_masters.push_back(local_masters[j]);
      flattened_coeffs.push_back(local_coeffs[j]);
    }
  }

  // Loop over all masters and move contributions that are outside the local
  // cell into b. We remove contributions in local in-out element vector
  for (std::int32_t slv_idx = 0; slv_idx < flattened_slaves.size(); ++slv_idx)
  {
    const std::int32_t& local_index = flattened_slaves[slv_idx];
    const std::int32_t& master = flattened_masters[slv_idx];
    const T& coeff = flattened_coeffs[slv_idx];

    for (std::int32_t i = 0; i < num_dofs; ++i)
    {
      for (int j = 0; j < bs; ++j)
      {
        b[master] += coeff * b_local_org[i * bs + j];
        b_local[i * bs + j] = 0;
      }
    }
  }
}

/// Assemble an integration kernel over a set of active entities, described
/// through into vector of type T, and apply the multipoint constraint
/// @param[in, out] b The vector to assemble into
/// @param[in] active_entities The set of active entities.
/// @param[in] dofmap The dofmap
/// @param[in] bs The block size of the dofmap
/// @param[in] coeffs The packed coefficients for all cells
/// @param[in] constants The pack constants
/// @param[in] cell_info The cell permutation info
/// @param[in] mpc The multipoint constraint
/// @param[in] fetch_cells Function that fetches the cell index for an entity
/// in active_entities
/// @param[in] assemble_local_element_matrix Function f(be, index) that
/// assembles into a local element matrix for a given entity
template <typename T, typename E_DESC>
void _assemble_entities_impl(
    xtl::span<T> b, const dolfinx::mesh::Mesh& mesh,
    const std::vector<E_DESC>& active_entities,
    const dolfinx::graph::AdjacencyList<std::int32_t>& dofmap, int bs,
    const xtl::span<const T> coeffs, int cstride,
    const std::vector<T>& constants,
    const xtl::span<const std::uint32_t>& cell_info,
    const std::shared_ptr<const dolfinx_mpc::MultiPointConstraint<T>>& mpc,
    const std::function<const std::int32_t(const E_DESC&)> fetch_cells,
    const std::function<void(xtl::span<T>, E_DESC)>
        assemble_local_element_vector)
{
  // Get MPC data
  const std::shared_ptr<const dolfinx::graph::AdjacencyList<std::int32_t>>
      masters = mpc->masters_local();
  const std::shared_ptr<const dolfinx::graph::AdjacencyList<T>> coefficients
      = mpc->coeffs();
  const xtl::span<const std::int32_t> slaves = mpc->slaves();

  xtl::span<const std::int32_t> slave_cells = mpc->slave_cells();
  const std::shared_ptr<const dolfinx::graph::AdjacencyList<std::int32_t>>
      cell_to_slaves = mpc->cell_to_slaves();

  // Compute local indices for slave cells
  std::vector<bool> is_slave_entity
      = std::vector<bool>(active_entities.size(), false);
  std::vector<std::int32_t> slave_cell_index
      = std::vector<std::int32_t>(active_entities.size(), -1);
  for (std::int32_t i = 0; i < active_entities.size(); ++i)
  {
    const std::int32_t cell = fetch_cells(active_entities[i]);
    for (std::int32_t j = 0; j < slave_cells.size(); ++j)
      if (slave_cells[j] == cell)
      {
        is_slave_entity[i] = true;
        slave_cell_index[i] = j;
        break;
      }
  }

  // NOTE: Assertion that all links have the same size (no P refinement)
  const int num_dofs = dofmap.links(0).size();
  std::vector<T> be(bs * num_dofs);
  const xtl::span<T> _be(be);
  std::vector<T> be_copy(bs * num_dofs);
  const xtl::span<T> _be_copy(be_copy);

  // Local vector to indicate which dof is slave
  std::vector<bool> is_slave(bs * num_dofs);

  // Assemble over all entities
  for (std::int32_t e = 0; e < active_entities.size(); ++e)
  {
    // Assemble into element vector
    assemble_local_element_vector(_be, active_entities[e]);

    const std::int32_t cell = fetch_cells(active_entities[e]);
    auto dofs = dofmap.links(cell);

    // Modify local element matrix if entity is connected to a slave cell
    if (is_slave_entity[e])
    {
      // Find local position of every slave in cell
      xtl::span<const int32_t> slave_indices
          = cell_to_slaves->links(slave_cell_index[e]);
      std::vector<std::int32_t> local_indices(slave_indices.size());
      for (std::int32_t i = 0; i < slave_indices.size(); ++i)
      {
        bool found = false;
        for (std::int32_t j = 0; j < dofs.size(); ++j)
        {
          for (std::int32_t k = 0; k < bs; ++k)
          {
            if (bs * dofs[j] + k == slaves[slave_indices[i]])
            {
              local_indices[i] = bs * j + k;
              break;
            }
          }
          if (found)
            break;
        }
      }
      // Modify element vector for MPC and insert into b for non-local
      // contributions
      std::copy(be.begin(), be.end(), be_copy.begin());
      modify_mpc_vec<T>(b, _be, _be_copy, num_dofs, bs, slave_indices,
                        local_indices, masters, coefficients);
    }
    // Add local contribution to b
    for (int i = 0; i < num_dofs; ++i)
      for (int k = 0; k < bs; ++k)
        b[bs * dofs[i] + k] += be[bs * i + k];
  }
}

template <typename T>
void _assemble_vector(
    xtl::span<T> b, const dolfinx::fem::Form<T>& L,
    const std::shared_ptr<const dolfinx_mpc::MultiPointConstraint<T>>& mpc)
{

  std::shared_ptr<const dolfinx::mesh::Mesh> mesh = L.mesh();
  assert(mesh);
  const int tdim = mesh->topology().dim();

  // Get dofmap data
  std::shared_ptr<const dolfinx::fem::DofMap> dofmap
      = L.function_spaces().at(0)->dofmap();
  assert(dofmap);
  const dolfinx::graph::AdjacencyList<std::int32_t>& dofs = dofmap->list();
  const int bs = dofmap->bs();

  const int num_dofs = dofs.links(0).size();
  const std::uint32_t ndim = bs * num_dofs;

  // Prepare constants & coefficients
  const std::vector<T> constants = pack_constants(L);
  const auto coeffs = dolfinx::fem::pack_coefficients(L);

  // Prepare cell geometry
  const dolfinx::graph::AdjacencyList<std::int32_t>& x_dofmap
      = mesh->geometry().dofmap();
  const xt::xtensor<double, 2>& x_g = mesh->geometry().x();

  // Prepare dof tranformation data
  std::shared_ptr<const dolfinx::fem::FiniteElement> element
      = L.function_spaces().at(0)->element();
  const std::function<void(const xtl::span<T>&,
                           const xtl::span<const std::uint32_t>&, std::int32_t,
                           int)>
      dof_transform = element->get_dof_transformation_function<T>();
  const bool needs_transformation_data
      = element->needs_dof_transformations() or L.needs_facet_permutations();
  xtl::span<const std::uint32_t> cell_info;
  if (needs_transformation_data)
  {
    mesh->topology_mutable().create_entity_permutations();
    cell_info = xtl::span(mesh->topology().get_cell_permutation_info());
  }

  if (L.num_integrals(dolfinx::fem::IntegralType::cell) > 0)
  {
    const auto fetch_cells = [&](const std::int32_t& entity) { return entity; };
    for (int i : L.integral_ids(dolfinx::fem::IntegralType::cell))
    {
      const auto& fn = L.kernel(dolfinx::fem::IntegralType::cell, i);

      /// Assemble local cell kernels into a vector
      /// @param[in] be The local element vector
      /// @param[in] cell The cell index
      const auto assemble_local_cell_vector
          = [&](xtl::span<T> be, std::int32_t cell)
      {
        // Fetch the coordiantes of the cell
        const xtl::span<const std::int32_t> x_dofs = x_dofmap.links(cell);
        const xt::xarray<double> coordinate_dofs(
            xt::view(x_g, xt::keep(x_dofs), xt::all()));

        // Tabulate tensor
        std::fill(be.data(), be.data() + be.size(), 0);
        fn(be.data(), coeffs.first.data() + cell * coeffs.second,
           constants.data(), coordinate_dofs.data(), nullptr, nullptr);

        // Apply any required transformations
        dof_transform(be, cell_info, cell, 1);
      };

      // Assemble over all active cells
      const std::vector<std::int32_t>& active_cells = L.cell_domains(i);
      _assemble_entities_impl<T, std::int32_t>(
          b, *mesh, active_cells, dofs, bs, coeffs.first, coeffs.second,
          constants, cell_info, mpc, fetch_cells, assemble_local_cell_vector);
    }
  }
}
} // namespace
//-----------------------------------------------------------------------------

void dolfinx_mpc::assemble_vector(
    xtl::span<double> b, const dolfinx::fem::Form<double>& L,
    const std::shared_ptr<const dolfinx_mpc::MultiPointConstraint<double>>& mpc)
{
  _assemble_vector<double>(b, L, mpc);
}
//-----------------------------------------------------------------------------

File: python/dolfinx_mpc/numba/assemble_vector.py
# Copyright (C) 2021 Jørgen S. Dokken
#
# This file is part of DOLFINX_MPC
#
# SPDX-License-Identifier:    MIT

from typing import Tuple

import dolfinx
import dolfinx.common
import dolfinx.log
import numpy
import ufl
from dolfinx_mpc.multipointconstraint import MultiPointConstraint
from petsc4py import PETSc

import numba

from .helpers import extract_slave_cells, pack_slave_facet_info
from .numba_setup import initialize_petsc

Timer = dolfinx.common.Timer

ffi, _ = initialize_petsc()


def assemble_vector(form: ufl.form.Form, constraint: MultiPointConstraint, b: PETSc.Vec = None,
                    form_compiler_parameters={}, jit_parameters={}) -> PETSc.Vec:
    """
    Assemble a linear form into vector b.

    Parameters
    ----------
    form
        The bilinear variational form
    constraint
        The multi point constraint
    b
        PETSc vector to assemble into (optional)
    form_compiler_parameters
        Parameters used in FFCx compilation of this form. Run `ffcx --help` at
        the commandline to see all available options. Takes priority over all
        other parameter values, except for `scalar_type` which is determined by
        DOLFINx.
    jit_parameters
        Parameters used in CFFI JIT compilation of C code generated by FFCx.
        See `python/dolfinx/jit.py` for all available parameters.
        Takes priority over all other parameter values.

    """

    dolfinx.log.log(dolfinx.log.LogLevel.INFO, "Assemble MPC vector")
    timer_vector = Timer("~MPC: Assemble vector (numba)")

    # Unpack Function space data
    V = form.arguments()[0].ufl_function_space()
    pos = V.mesh.geometry.dofmap.offsets
    x_dofs = V.mesh.geometry.dofmap.array
    x = V.mesh.geometry.x
    dofs = V.dofmap.list.array
    block_size = V.dofmap.index_map_bs

    # Data from multipointconstraint
    coefficients = constraint.coefficients()[0]
    masters_adj = constraint.masters
    c_to_s_adj = constraint.cell_to_slaves
    cell_to_slave = c_to_s_adj.array
    c_to_s_off = c_to_s_adj.offsets
    is_slave = constraint.is_slave
    mpc_data = (masters_adj.array, coefficients, masters_adj.offsets, cell_to_slave, c_to_s_off, is_slave)
    slave_cells = extract_slave_cells(c_to_s_off)

    # Get index map and ghost info
    if b is None:
        index_map = constraint.function_space.dofmap.index_map
        vector = dolfinx.cpp.la.create_vector(index_map, block_size)
    else:
        vector = b

    # If using DOLFINx complex build, scalar type in form_compiler parameters must be updated
    is_complex = numpy.issubdtype(PETSc.ScalarType, numpy.complexfloating)
    if is_complex:
        form_compiler_parameters["scalar_type"] = "double _Complex"

    # Compile ufc form for Python assembly
    ufc_form, _, _ = dolfinx.jit.ffcx_jit(V.mesh.mpi_comm(), form,
                                          form_compiler_parameters=form_compiler_parameters,
                                          jit_parameters=jit_parameters)

    # Pack constants and coefficients
    cpp_form = dolfinx.Form(form, form_compiler_parameters=form_compiler_parameters,
                            jit_parameters=jit_parameters)._cpp_object
    form_coeffs = dolfinx.cpp.fem.pack_coefficients(cpp_form)
    form_consts = dolfinx.cpp.fem.pack_constants(cpp_form)
    tdim = V.mesh.topology.dim
    num_dofs_per_element = V.dofmap.dof_layout.num_dofs

    # Assemble vector with all entries
    dolfinx.cpp.fem.assemble_vector(vector.array_w, cpp_form, form_consts, form_coeffs)

    # Check if we need facet permutations
    # FIXME: access apply_dof_transformations here
    e0 = cpp_form.function_spaces[0].element
    needs_transformation_data = e0.needs_dof_transformations or cpp_form.needs_facet_permutations
    cell_perms = numpy.array([], dtype=numpy.uint32)
    if needs_transformation_data:
        V.mesh.topology.create_entity_permutations()
        cell_perms = V.mesh.topology.get_cell_permutation_info()
    if e0.needs_dof_transformations:
        raise NotImplementedError("Dof transformations not implemented")
    # Assemble over cells
    subdomain_ids = cpp_form.integral_ids(dolfinx.fem.IntegralType.cell)
    num_cell_integrals = len(subdomain_ids)
    nptype = "complex128" if is_complex else "float64"
    if num_cell_integrals > 0:
        V.mesh.topology.create_entity_permutations()
        for i, id in enumerate(subdomain_ids):
            cell_kernel = getattr(ufc_form.integrals(dolfinx.fem.IntegralType.cell)[i], f"tabulate_tensor_{nptype}")
            active_cells = cpp_form.domains(dolfinx.fem.IntegralType.cell, id)
            coeffs_i = form_coeffs[(dolfinx.fem.IntegralType.cell, id)]
            with vector.localForm() as b:
                assemble_cells(numpy.asarray(b), cell_kernel, active_cells[numpy.isin(active_cells, slave_cells)],
                               (pos, x_dofs, x), coeffs_i, form_consts,
                               cell_perms, dofs, block_size, num_dofs_per_element, mpc_data)

    # Assemble exterior facet integrals
    subdomain_ids = cpp_form.integral_ids(dolfinx.fem.IntegralType.exterior_facet)
    num_exterior_integrals = len(subdomain_ids)
    if num_exterior_integrals > 0:
        V.mesh.topology.create_entities(tdim - 1)
        V.mesh.topology.create_connectivity(tdim - 1, tdim)
        # Get facet permutations if required
        facet_perms = numpy.array([], dtype=numpy.uint8)
        if cpp_form.needs_facet_permutations:
            facet_perms = V.mesh.topology.get_facet_permutations()
        perm = (cell_perms, cpp_form.needs_facet_permutations, facet_perms)
        for i, id in enumerate(subdomain_ids):
            facet_kernel = getattr(ufc_form.integrals(dolfinx.fem.IntegralType.exterior_facet)[i],
                                   f"tabulate_tensor_{nptype}")
            coeffs_i = form_coeffs[(dolfinx.fem.IntegralType.exterior_facet, id)]
            facets = cpp_form.domains(dolfinx.fem.IntegralType.exterior_facet, id)
            facet_info = pack_slave_facet_info(facets, slave_cells)
            num_facets_per_cell = len(V.mesh.topology.connectivity(tdim, tdim - 1).links(0))
            with vector.localForm() as b:
                assemble_exterior_slave_facets(numpy.asarray(b), facet_kernel, facet_info, (pos, x_dofs, x),
                                               coeffs_i, form_consts, perm,
                                               dofs, block_size, num_dofs_per_element, mpc_data, num_facets_per_cell)
    timer_vector.stop()
    return vector


@numba.njit
def assemble_cells(b: 'numpy.ndarray[PETSc.ScalarType]',
                   kernel: ffi.CData, active_cells: 'numpy.ndarray[numpy.int32]',
                   mesh: Tuple['numpy.ndarray[numpy.int32]', 'numpy.ndarray[numpy.int32]',
                               'numpy.ndarray[numpy.float64]'],
                   coeffs: 'numpy.ndarray[PETSc.ScalarType]',
                   constants: 'numpy.ndarray[PETSc.ScalarType]',
                   permutation_info: 'numpy.ndarray[numpy.uint32]',
                   dofmap: 'numpy.ndarray[numpy.int32]',
                   block_size: int,
                   num_dofs_per_element: int,
                   mpc: Tuple['numpy.ndarray[numpy.int32]', 'numpy.ndarray[PETSc.ScalarType]',
                              'numpy.ndarray[numpy.int32]', 'numpy.ndarray[numpy.int32]',
                              'numpy.ndarray[numpy.int32]', 'numpy.ndarray[numpy.int32]']):
    """Assemble additional MPC contributions for cell integrals"""
    ffi_fb = ffi.from_buffer

    # Empty arrays mimicking Nullpointers
    facet_index = numpy.zeros(0, dtype=numpy.int32)
    facet_perm = numpy.zeros(0, dtype=numpy.uint8)

    # Unpack mesh data
    pos, x_dofmap, x = mesh

    # NOTE: All cells are assumed to be of the same type
    geometry = numpy.zeros((pos[1] - pos[0], 3))
    b_local = numpy.zeros(block_size * num_dofs_per_element, dtype=PETSc.ScalarType)

    for cell_index in active_cells:
        num_vertices = pos[cell_index + 1] - pos[cell_index]
        cell = pos[cell_index]

        # Compute mesh geometry for cell
        geometry[:, :] = x[x_dofmap[cell:cell + num_vertices]]

        # Assemble local element vector
        b_local.fill(0.0)
        kernel(ffi_fb(b_local), ffi_fb(coeffs[cell_index, :]),
               ffi_fb(constants), ffi_fb(geometry), ffi_fb(facet_index),
               ffi_fb(facet_perm))
        # FIXME: Here we need to add the apply_dof_transformation function

        # Modify global vector and local cell contributions
        b_local_copy = b_local.copy()
        modify_mpc_contributions(b, cell_index, b_local, b_local_copy, mpc, dofmap,
                                 block_size, num_dofs_per_element)
        for j in range(num_dofs_per_element):
            for k in range(block_size):
                position = dofmap[num_dofs_per_element * cell_index + j] * block_size + k
                b[position] += (b_local[j * block_size + k] - b_local_copy[j * block_size + k])


@numba.njit
def assemble_exterior_slave_facets(b: 'numpy.ndarray[PETSc.ScalarType]',
                                   kernel: ffi.CData,
                                   facet_info: 'numpy.ndarray[numpy.int32]',
                                   mesh: Tuple['numpy.ndarray[numpy.int32]', 'numpy.ndarray[numpy.int32]',
                                               'numpy.ndarray[numpy.float64]'],
                                   coeffs: 'numpy.ndarray[PETSc.ScalarType]',
                                   constants: 'numpy.ndarray[PETSc.ScalarType]',
                                   permutation_info: 'numpy.ndarray[numpy.uint32]',
                                   dofmap: 'numpy.ndarray[numpy.int32]',
                                   block_size: int,
                                   num_dofs_per_element: int,
                                   mpc: Tuple['numpy.ndarray[numpy.int32]', 'numpy.ndarray[PETSc.ScalarType]',
                                              'numpy.ndarray[numpy.int32]', 'numpy.ndarray[numpy.int32]',
                                              'numpy.ndarray[numpy.int32]', 'numpy.ndarray[numpy.int32]'],
                                   num_facets_per_cell: int):
    """Assemble additional MPC contributions for facets"""
    ffi_fb = ffi.from_buffer

    # Unpack facet permutation info
    cell_perms, needs_facet_perm, facet_perms = permutation_info
    facet_index = numpy.zeros(1, dtype=numpy.int32)
    facet_perm = numpy.zeros(1, dtype=numpy.uint8)

    # Unpack mesh data
    pos, x_dofmap, x = mesh

    geometry = numpy.zeros((pos[1] - pos[0], 3))
    b_local = numpy.zeros(block_size * num_dofs_per_element, dtype=PETSc.ScalarType)
    for i in range(facet_info.shape[0]):
        # Extract cell index (local to process) and facet index (local to cell) for kernel
        cell_index, local_facet = facet_info[i]
        facet_index[0] = local_facet

        # Extract cell geometry
        cell = pos[cell_index]
        num_vertices = pos[cell_index + 1] - pos[cell_index]
        geometry[:, :] = x[x_dofmap[cell:cell + num_vertices]]

        # Compute local facet kernel
        b_local.fill(0.0)
        if needs_facet_perm:
            facet_perm[0] = facet_perms[cell_index * num_facets_per_cell + local_facet]
        kernel(ffi_fb(b_local), ffi_fb(coeffs[cell_index, :]), ffi_fb(constants), ffi_fb(geometry),
               ffi_fb(facet_index), ffi_fb(facet_perm))
        # FIXME: Here we need to add the apply_dof_transformation

        # Modify local contributions and add global MPC contributions
        b_local_copy = b_local.copy()
        modify_mpc_contributions(b, cell_index, b_local, b_local_copy,
                                 mpc, dofmap, block_size, num_dofs_per_element)
        for j in range(num_dofs_per_element):
            for k in range(block_size):
                position = dofmap[num_dofs_per_element * cell_index + j] * block_size + k
                b[position] += (b_local[j * block_size + k] - b_local_copy[j * block_size + k])


@numba.njit(cache=True)
def modify_mpc_contributions(b: 'numpy.ndarray[PETSc.ScalarType]', cell_index: int,
                             b_local: 'numpy.ndarray[PETSc.ScalarType]',
                             b_copy: 'numpy.ndarray[PETSc.ScalarType]',
                             mpc: Tuple['numpy.ndarray[numpy.int32]', 'numpy.ndarray[PETSc.ScalarType]',
                                        'numpy.ndarray[numpy.int32]', 'numpy.ndarray[numpy.int32]',
                                        'numpy.ndarray[numpy.int32]', 'numpy.ndarray[numpy.int32]'],
                             dofmap: 'numpy.ndarray[numpy.int32]',
                             block_size: int,
                             num_dofs_per_element: int):
    """
    Modify local entries of b_local with MPC info and add modified
    entries to global vector b.
    """

    # Unwrap MPC data
    masters, coefficients, offsets, cell_to_slave, cell_to_slave_offset, is_slave = mpc

    # Determine which slaves are in this cell,
    # and which global index they have in 1D arrays
    cell_slaves = cell_to_slave[cell_to_slave_offset[cell_index]:
                                cell_to_slave_offset[cell_index + 1]]

    # Get local index of slaves in cell
    cell_blocks = dofmap[num_dofs_per_element * cell_index:
                         num_dofs_per_element * cell_index + num_dofs_per_element]
    local_index = numpy.empty(len(cell_slaves), dtype=numpy.int32)
    for i in range(num_dofs_per_element):
        for j in range(block_size):
            dof = cell_blocks[i] * block_size + j
            if is_slave[dof]:
                location = numpy.flatnonzero(cell_slaves == dof)[0]
                local_index[location] = i * block_size + j

    # Move contribution from each slave to the corresponding master dof
    # and zero out local b
    for local, slave in zip(local_index, cell_slaves):
        cell_masters = masters[offsets[slave]: offsets[slave + 1]]
        cell_coeffs = coefficients[offsets[slave]: offsets[slave + 1]]
        for m0, c0 in zip(cell_masters, cell_coeffs):
            b[m0] += c0 * b_copy[local]
            b_local[local] = 0
